content/schema-evolution-field-operations/index.md

```diff
@@ -0,0 +1,194 @@
+of a field).
+
+Note that you can find the [glossary](#glossary) of concepts used throughout
+this RFC at the end of it.
+
+Note: The pseudocode in this RFC resembles [Elm](http://elm-lang.org) but it
+is not intended to be valid.
+
+
+## Motivation
+
+Currently, the Registers Data Model has no way to evolve the data structure
+described in the schema and used in data items. This means a Register has to
+be designed upfront with the right fields forever given it is an append-only
+log. We need a way to be able to evolve the schema definition and a way to
+reflect such evolution on existing and new data.
+
+
+## Explanation
+
+The proposal is to allow adding new field definitions to the schema. As a
+result new items added to the Register will be validated against the new
+schema. To implement this behaviour this RFC defines the mechanism to [add a
+field to the schema](#field-addition) and the mechanism to [add an item to the
+log](#item-addition).
+
+Note: The explanation tries to avoid being prescriptive in how the Register is
+implemented and because of that there is no explicit reference to entries and
+how they participate in either mechanism.
+
+
+### <span id="field-addition">Field addition</span>
+
+The addition of a field, [`Schema.add`](#schema-add-function), adds a
+[field](#field) to the [schema](#schema).
+
+
+##### Example: Add a field to the schema
+
+```elm
+s_0 = Schema [("x", Number)]
+s_1 = Schema.add s_0 ("y", Number) -- s_1 == Schema [("x", Number), ("y", Number)]
+```
+
+
+### <span id="item-addition">Item addition</span>
+
+The item addition operation, [`Item.add`](#item-add-function), adds the item
+if [`Item.validate`](#item-validate-function) succeeds. This RFC doesn't
+change its behaviour, the only thing to keep in mind is that the validation
+function checks against the current schema.
+
+The addition of an item steps are:
+
+1. Validate `a` against the current schema `s`: `Item.validate s a`.
+    * If the item is invalid, reject the operation.
+2. Canonicalise the item `a`: `a_c = Item.canonicalise a`.
+3. Add item `a_c` to the Register.
+
+
+### REST API Implications
+
+The REST API has no changes given that it is a read-only API. From the **API
+consumer** perspective, nothing changes but let me make it explicit here.
+
+One cornerstone assumption in the Registers' API is that a consumer is
+expected to handle graciously any Record/Item response that doesn't match
+exactly their knowledge of the schema.
+
+Scenario:
+
+Give a consumer that knows about a schema such as `Schema [("x", Number),
+("y", Number), ("z", String)]` where `x` is the primary field, such consumer
+is expected to handle reponses from the API such as:
+
+* **All known fields are present**
+
+  Given `{"x": 1, "y": 2, "z": "foo"}`, keep `{"x": 1, "y": 2, "z": "foo"}`.
+
+* **Some known fields are present**
+
+  Given `{"x": 1, "y": 2}`, keep `{"x": 1, "y": 2}` and ignore the fact that
+  `z` is missing.
+
+* **Some unknown fields present**
```

Comment:
I agree that this will work for field addition - that "a consumer is expected to handle graciously a response that doesn't match their knowledge of the schema". However, especially for this last case, I bet there are some people that haven't integrated in this way, and are simply consuming what we give them in the item. Do you feel that this is a bit "new" and we need to make it really obvious in our API documentation that we expect this?

Reply:
For my conversations with Paul, this is the intended way since the beginning. That said, I would like the API docs to be much more clear about this expectations.

Now, if someone prefers accepting any field found in an item because they don't use the schema and are happy to handle data in a different way, it's ok. The point is that they shouldn't get an exception because of this.

Reply:
Given that we don't know for sure all current implementations, I'm happy to start by putting a clear explanation about expectations in the API docs (sooner rather than latter) and put a expected data for when this will be used in our end.